torch/csrc/jit/mobile/nnc/context.cpp

```diff
@@ -0,0 +1,252 @@
+
+  // input_specs_
+  for (const auto& input_value : dict.at("input_specs").toTuple()->elements()) {
+    input_specs_.emplace_back(input_value);
+  }
+
+  // output_specs_
+  for (const auto& output_value :
+       dict.at("output_specs").toTuple()->elements()) {
+    output_specs_.emplace_back(output_value);
+  }
+
+  // memory_plan_
+  memory_plan_ = MemoryPlan(dict.at("memory_plan"));
+}
+
+c10::IValue Function::serialize() const {
+  c10::Dict<c10::IValue, c10::IValue> dict(
+      at::StringType::get(), at::AnyType::get());
+
+  dict.insert("name", name_.qualifiedName());
+  dict.insert("nnc_kernel_id", nnc_kernel_id_);
+  dict.insert("parameters", parameters_);
+
+  // input_specs_
+  std::vector<c10::IValue> inputs;
+  for (const auto& input_spec : input_specs_) {
+    inputs.emplace_back(input_spec.serialize());
+  }
+  dict.insert("input_specs", Tup(std::move(inputs)));
+
+  // output_specs_
+  std::vector<c10::IValue> outputs;
+  for (const auto& output_spec : output_specs_) {
+    outputs.emplace_back(output_spec.serialize());
+  }
+  dict.insert("output_specs", Tup(std::move(outputs)));
+
+  // memory_plan_
+  dict.insert("memory_plan", memory_plan_.serialize());
+  return dict;
+}
+
+void Function::init_execution_state() const {
+  if (execution_state_.get() != nullptr) {
+    return;
+  }
+
+  ExecutionState state;
+  memory_plan_.allocate(&state);
+
+  // The arguments vector consists of 4 sections: inputs, outputs, parameters
+  // and buffers.
+  auto input_args = input_specs_.size();
+  auto output_args = output_specs_.size();
+  auto param_args = parameters_.size();
+  auto buffer_args = state.preallocations_.size();
+
+  auto& arguments = state.arguments_;
+  arguments.reserve(input_args + output_args + param_args + buffer_args);
+
+  // Keep empty slots to fill in inputs/outputs pointers at execution time.
+  arguments.resize(input_args + output_args);
+
+  // Fill in parameter pointers.
+  for (const auto& param : parameters_) {
+    arguments.emplace_back(param.data_ptr());
+  }
+
+  // Fill in preallocated buffer pointers.
+  for (const auto& preallocation : state.preallocations_) {
+    arguments.emplace_back(preallocation.get());
+  }
+
+  execution_state_ = std::make_unique<ExecutionState>(std::move(state));
+}
+
+c10::impl::GenericList Function::run(
+    const c10::impl::GenericList& inputs) const {
+  TORCH_CHECK(
+      registry::has_nnc_kernel(nnc_kernel_id_),
+      "Cannot find NNC kernel: ",
+      nnc_kernel_id_);
+
+  init_execution_state();
+
+  std::vector<void*>& args = execution_state_->arguments_;
+
+  // Fill in input tensors.
+  TORCH_CHECK(
+      input_specs_.size() == inputs.size(),
+      "Input size doesn't match the spec, expect: ",
+      input_specs_.size(),
+      " actual: ",
+      inputs.size());
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    const c10::IValue& input = inputs[i];
+    const auto& input_tensor = input.toTensor();
+    TORCH_CHECK(
```

Comment:
So all these checks can also be using DelegatedBackendException with appropriate debug handle? Not for now but when we integrate debug handles.

Reply:
Yes we can - but the model hasn't really started running, guess the most useful context is probably just the pointer to the compiled Module + Method?
This "run" function is called from the generated TorchScript wrapper code (executed by the standard lite-interpreter), where it should have the Module/Method sources as well. So it could also catch and re-throw with the debug-handle saved there.

Reply:
Makes sense.

Reply:
Currently we assume the call succeeds: self.__backend.execute(self.__handles["$method"], typed_inputs)

Or else the backend triggers the DelegatedBackendException with the right source data.

@kimishpatel @iseeyuan  Can you please take a look to make sure that even if the backend fails with another exception we fail with some useful information like you describe above?
@kimishpatel  This should probably be part of your PR where you introduce the DelegatedBackendException

Reply:
Maybe for backend we should use the other macro for delegate exceptions `TORCH_DELEGATED_BACKEND_THROW`? You can use the debug handle corresponding to the first node of the graph or just use sentinel value of -1. However the former would be better such that you can at least get some stack trace, if say you lower only some submodule to nnc. So generally speaking moving to DelegatedBackendException is what we should do even if the information is sparse so that everything flows through it. What do you think @raziel?